File: 28.implement-str-str.java
/*
 * @lc app=leetcode id=28 lang=java
 *
 * [28] Implement strStr()
 *
 * https://leetcode.com/problems/implement-strstr/description/
 *
 * algorithms
 * Easy (34.05%)
 * Likes:    1412
 * Dislikes: 1772
 * Total Accepted:    622.2K
 * Total Submissions: 1.8M
 * Testcase Example:  '"hello"\n"ll"'
 *
 * Implement strStr().
 *
 * Return the index of the first occurrence of needle in haystack, or -1 if
 * needle is not part of haystack.
 *
 * Example 1:
 *
 *
 * Input: haystack = "hello", needle = "ll"
 * Output: 2
 *
 *
 * Example 2:
 *
 *
 * Input: haystack = "aaaaa", needle = "bba"
 * Output: -1
 *
 *
 * Clarification:
 *
 * What should we return when needle is an empty string? This is a great
 * question to ask during an interview.
 *
 * For the purpose of this problem, we will return 0 when needle is an empty
 * string. This is consistent to C's strstr() and Java's indexOf().
 *
 */

// @lc code=start
class Solution {
    public int strStr(String haystack, String needle) {
        if (needle.isEmpty()) return 0;
        int len = needle.length();
        for (int i = 0; i <= haystack.length() - len; i++) {
            if (needle.equals(haystack.substring(i, i+len)))
                return i;
        }
        return -1;
    }
}
// @lc code=end

File: 28.implement-str-str.cpp
/*
 * @lc app=leetcode id=28 lang=cpp
 *
 * [28] Implement strStr()
 */

// @lc code=start
#include <string>
using std::string;
class Solution {
public:
    int strStr(string haystack, string needle) {
        if (!needle.size()) return 0;
        if (haystack.size() < needle.size()) return -1;
        for (int i = 0; i != haystack.size() - needle.size() + 1; ++i) {
            if (isSameString(haystack, needle, i))
                return i;
        }
        return -1;
    } 
private:
    bool isSameString(string& haystack, string& needle, int offset){
        for(int i = 0; i != needle.size(); ++i)
            if (haystack[i+offset] != needle[i])
                return false;
        return true;
    }
};
// @lc code=end

